### PR TITLE
Add ExprSentCommands and EvtPlayerCommandSend

### PR DESCRIPTION
**Pull request by sovdeeth**
### Description

Adds functionality to listen to and modify PlayerCommandSendEvents. 
The sent commands can be modified, but not added to. Modifications are reflected in the tab completes available to the player. This has no effect on the commands they can actually run.

Would like opinions on the syntax for the event/expresssion.
Skbee uses `command map` to reference the collection of commands, which could be useful for ease of conversion.

---
**Target Minecraft Versions:** any 
**Requirements:** 
**Related Issues:** #3356

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.skriptlang.skript.lang.structure.Structure;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+@Name("Sent Commands")
+@Description({
+	"The commands that will be sent to the player in a <a href='events.html#send_commands_to_player'>send commands to player event</a>.",
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:46]: Reminder: docs reference should reflect new name if renamed

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "the sent commands";
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:149]:
```suggestion
		return "the server sent commands list";
```

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Nice PR! just few things

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+@Examples({
+		"on send commands to player:",
+			"\tset sent commands to sent commands where [input does not contain \":\"]",
+			"\tadd \"fly\" to sent commands"
+})
+@Since("INSERT VERSION")
+public class EvtPlayerCommandSend extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Send Commands to Player", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] commands to player");
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:53]: I think whoever came up with that is incredibly smart and handsome

[the remark this answers is not in the document]
Comment (Moderocky) [reply src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:53]: I don't know what to think about these patterns, maybe something like on sending command list sounds better?

[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:53]: My plan was `on send[ing] [of] [server] command[s] list`

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		"Called when the server sends a list of commands to the player. The sent commands can be modified via the <a href='expressions.html#ExprSentCommands'>sent commands expression</a>.",
+		"Modifications will affect what commands show up for the player to tab complete. They will not affect what commands the player can actually run.",
+		"Adding new commands to the list is illegal behavior and will be ignored."
+})
+@Examples({
+		"on send commands to player:",
+			"\tset sent commands to sent commands where [input does not contain \":\"]",
+			"\tadd \"fly\" to sent commands"
+})
+@Since("INSERT VERSION")
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:49]: i should know this by now lol

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.bukkit.event.player.PlayerCommandSendEvent;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+public class EvtPlayerCommandSend extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Send Command List", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] [of] [server] command[s] list ")
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:37]:
```suggestion
		Skript.registerEvent("Send Command List", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] [of] [server] command[s] list")
```

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+				)
+				.examples(
+					"on send command list:",
+						"\tset command list to command list where [input does not contain \":\"]",
+						"\tremove \"help\" from command list"
+				)
+				.since("INSERT VERSION");
+	}
+
+	Collection<String> originalCommands = new ArrayList<>();
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:52]: private?

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		Skript.registerExpression(ExprSentCommands.class, String.class, ExpressionType.SIMPLE, "[the] [sent] [server] command[s] list");
+	}
+
+	private EvtPlayerCommandSend parent;
+	private Kleenean delay;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!getParser().isCurrentEvent(PlayerCommandSendEvent.class))
+			return false;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:69]: Should send some error to the user like `You can't use Expr outside Evt event`

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+		Structure structure = getParser().getCurrentStructure();
+		if (!(structure instanceof EvtPlayerCommandSend))
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:71]: Can this actually happen while current event is set to `PlayerCommandSendEvent` or is it just for safety if so maybe we should assert instead?

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java (hunk 8 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:71]: As of now, I don't think it can happen, but it could if some other skript event also triggered off of PlayerCommandSendEvent I think.

**Review by Moderocky — APPROVED** [review]
Looks good :)

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	}
+
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "the sent server command list";
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:157]:
```suggestion
		return "sent server command list";
```

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+				break;
+			case SET:
+				// remove all completely new commands, as adding new commands to the commands collection is illegal behaviour
+				List<String> newCommands = new ArrayList<>(deltaCommands);
+				newCommands.removeAll(parent.getOriginalCommands());
+				deltaCommands.removeAll(newCommands);
+				commands.clear();
+				commands.addAll(deltaCommands);
+				break;
+			case RESET:
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:134]: Because of the other comments, resetting of these values won't be needed as another plugin or script may have already modified the values, so the reset values are not true.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+	@Override
+	@Nullable
+	protected String[] get(Event event) {
+		if (delay != Kleenean.FALSE) {
+			Skript.error("Can't change the command list after the event has already passed");
+			return null;
+		}
+		if (!(event instanceof PlayerCommandSendEvent)) {
+			Skript.error("The 'command list' expression can only be used in a 'send command list' event");
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:86]: Whoops, must have put it in the wrong spot! Thanks for the catch.

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+	public ImmutableCollection<String> getOriginalCommands() {
+		return ImmutableList.copyOf(originalCommands);
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "send server command list";
+	}
+
```
[the remark this answers is not in the document]
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:74]: This would prevent a user from resetting the command list after modifying it.
```
set command list to "xyz"
reset command list
```
would no longer work properly. 

The current implementation should not affect whether or not other scripts/plugins can modify the value. The original command list is set only when check is called, right before the event is executed. This means any manipulation prior to the event will work fine, and since this expression does not allow delays, nothing outside of the event can modify it until the skript event is done.

To sum up, using the change method only would cause unexpected behavior due to the state at the beginning of execution not being stored, and the current method doesn't interfere with other plugin or script modifications.

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+			return false;
+		}
+
+		Structure structure = getParser().getCurrentStructure();
+		if (!(structure instanceof EvtPlayerCommandSend)) {
+			Skript.error("The 'command list' expression can only be used in a 'send command list' event");
+			return false;
+		}
+
+		if (isDelayed.isFalse()) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:78]: Should this be negated?

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case REMOVE:
+			case REMOVE_ALL:
+			case DELETE:
+			case SET:
+			case RESET:
+				return new Class[]{String[].class};
+			case ADD:
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: Why isn't ADD allowed in this changer? You can mutably edit the list of the event.

**Review by TheLimeGlass — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java (hunk 13 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:78]: of course, silly mistake, thanks

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java (hunk 14 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: Because adding new commands to the list is illegal behavior, I thought it best to explicitly discouraged adding. You can get around it if you have a specific need (re-adding removed values) by using the SET changer, but the better practice would be to a) only set it to what you want, or b) only remove what you want.

Comment (TheLimeGlass) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: The JavaDoc of the getCommands states it's mutable, which means it's designed to be modified.

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: Mutable, yes, adding new commands, no.
> It is not legal to add entries to this collection, only remove them. Behaviour of adding entries is undefined.

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	@Override
+	public Class<? extends String> getReturnType() {
+		return String.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "sent server command list";
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:159]:
```suggestion
		return "the sent server command list";
```

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case REMOVE:
+			case REMOVE_ALL:
+			case DELETE:
+			case SET:
+			case RESET:
+				return new Class[]{String[].class};
+			case ADD:
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: There is not a reason to explicitly define this case, right?

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+			return null;
+		return ((PlayerCommandSendEvent) event).getCommands().toArray(new String[0]);
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case REMOVE:
+			case REMOVE_ALL:
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:99]: REMOVE_ALL does not (should not) need to be supported here

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	}
+
+	private EvtPlayerCommandSend parent;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (!getParser().isCurrentEvent(PlayerCommandSendEvent.class)) {
+			Skript.error("The 'command list' expression can only be used in a 'send command list' event");
+			return false;
+		}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:70]: This check is redundant with the Structure check below it.

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+import org.bukkit.event.player.PlayerCommandSendEvent;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+public class EvtPlayerCommandSend extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Send Command List", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] [of] [server] command[s] list")
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:37]: Thoughts on support syntax like `on command list send`? Also `on sending of the server command list` (though that is a bit long)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java (hunk 16 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: No, but I like to be explicit about every option for a switch case, since it makes it easier to see what corresponds to what action, in my opinion. I can remove it if you'd rather `default` just handle it all.

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:104]: IMO it is better for default to handle it... you could maybe add a comment describing why ADD is not supported? :D

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+		return true;
+	}
+
+	public ImmutableCollection<String> getOriginalCommands() {
+		return ImmutableList.copyOf(originalCommands);
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "send server command list";
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:72]:
```suggestion
		return "sending of the server command list";
```

File: src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+import org.bukkit.event.player.PlayerCommandSendEvent;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+public class EvtPlayerCommandSend extends SkriptEvent {
+
+	static {
+		Skript.registerEvent("Send Command List", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] [of] [the] [server] command[s] list", "[server] command list send")
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/events/EvtPlayerCommandSend.java:37]:
```suggestion
		Skript.registerEvent("Send Command List", EvtPlayerCommandSend.class, PlayerCommandSendEvent.class, "send[ing] [of [the]] [server] command[s] list", "[server] command list send")
```
so that `on send the command list` isn't valid

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+				newCommands.removeAll(parent.getOriginalCommands());
+				deltaCommands.removeAll(newCommands);
+				commands.clear();
+				commands.addAll(deltaCommands);
+				break;
+			case RESET:
+				commands.clear();
+				commands.addAll(parent.getOriginalCommands());
+				break;
+			case ADD:
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:134]:
```suggestion
```

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
just about perfect :)

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprSentCommands.java [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		if (!(event instanceof PlayerCommandSendEvent))
+			return;
+
+		Collection<String> commands = ((PlayerCommandSendEvent) event).getCommands();
+
+		// short circuit if we're just clearing the list
+		if (mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE_ALL) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSentCommands.java:112]: REMOVE ALL check can be removed

**Review by APickledWalrus — APPROVED** [review]
